Fix Issue #19910: Changed InputMode for float editor to decimal

Ahmed-Bhouri opened the pull request:
## Overview

1. This PR fixes #19910.
2. This PR does the following: Changes the input mode from numeric to decimal so that the decimal point appears on the IOS keyboard. and updates input to accept the keyboard's default decimal point depending on the course language.

## Essential Checklist

- [x] The **PR title** starts with "Fix #bugnum: " or "Fix part of #bugnum: ...", followed by a short, clear summary of the changes.
- [x] I have followed the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request).
- [x] I have assigned the correct reviewers to this PR (or will leave a comment with the phrase "@{{reviewer_username}} PTAL" if I don't have permissions to assign reviewers directly).
- [x] The linter/Karma presubmit checks pass on my local machine, and my PR follows the [coding style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide)).
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)

## Proof that changes are correct

after the changes the comma finally appears on the keyboard while using the input:

![keyboard](https://github.com/oppia/oppia/assets/49922043/f07394f6-5b10-4133-beff-335221e9cb58)

https://github.com/oppia/oppia/assets/49922043/4f0da923-4f95-4ab4-9f99-ae744ac111a5

https://github.com/oppia/oppia/assets/49922043/cd4511cb-323f-4d86-b1d0-0b0b238627b1

https://github.com/oppia/oppia/assets/49922043/924d889f-ace5-432d-9718-7afefc00e8cc



## PR Pointers

- Never force push! If you do, your PR will be closed.
- To reply to reviewers, follow these instructions: https://github.com/oppia/oppia/wiki/Make-a-pull-request#step-5-address-review-comments-until-all-reviewers-approve
- Some e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).
- See the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers) for what code owners will expect.

Ahmed-Bhouri committed:
Change inputmode to decimal in float editor

Ahmed-Bhouri commented:
@seanlip PTAL

seanlip commented:
@Ahmed-Bhouri Please reassign me once the discussion on the issue is completed and we are clear on the path forward. Thanks!

Ahmed-Bhouri committed:
Merge remote-tracking branch 'upstream/develop' into inputmode-decimal

Ahmed-Bhouri committed:
Accept Dot/Comma in float editor

Ahmed-Bhouri commented:
@seanlip PTAL

I have finally solved the issue as we discussed.

Sorry for the delay ( I had issues with pushing my code that turned out to be caused by my setup ( This is my first PR and I am running my dev environment on a codespaces instance) )

seanlip commented:
@Ahmed-Bhouri Thanks -- could you please show the full flow, including submission of the answer? Ideally demonstrate it for both the "." and "," cases. Thanks!

(Also, just a question -- I'm not familiar with Codespaces, but does it run the pre-push hook? Just wanted to check since that can help avoid CI errors in the PR.)

Ahmed-Bhouri committed:
Change input value to match currentDecimalSeparator before Submit

Ahmed-Bhouri committed:
Merge remote-tracking branch 'upstream/develop' into inputmode-decimal

Ahmed-Bhouri commented:
@seanlip PTAL -- Thank you for mentioning the submission part ( I had to fix a small error there )
I think I am done now I added the video.

I only use codespace to create a separate environment from my machine in the cloud ( I ran into a lot of problems with my 16Gb of ram on a MacBook M1 ). This way I can use better resources.

seanlip commented:
@Ahmed-Bhouri Thanks, just a question. In your video I don't see the language/region change or the keyboard used to type "2.5". Can you display this too?

seanlip commented on core/templates/components/forms/schema-based-editors/schema-based-float-editor.component.ts:
```diff
@@ -247,17 +247,18 @@ export class SchemaBasedFloatEditorComponent
         // Parse number if the string is in proper format.
         this.localValue = this.numberConversionService.convertToEnglishDecimal(
-          this.localStringValue
+          this.localStringValue.replace(/\.|,/g, currentDecimalSeparator)
```
A question: Instead of doing localStringValue.replace() in multiple places, should we just be storing the right thing in there from the outset?

Ahmed-Bhouri replied:
Thank you for this!
I have improved it in the commit i am pushing now

Ahmed-Bhouri committed:
Merge remote-tracking branch 'upstream/develop' into inputmode-decimal

Ahmed-Bhouri committed:
cleaner code

Ahmed-Bhouri commented:
I addressed the note mentioned in the review. 
@seanlip PTAL
Thank you

seanlip commented on core/templates/components/forms/schema-based-editors/schema-based-float-editor.component.ts:
```diff
@@ -247,18 +247,25 @@ export class SchemaBasedFloatEditorComponent
     } else {
       // Make sure number is in a correct format.
+      let currentDecimalSeparator = this.getCurrentDecimalSeparator();
+      let formattedDtringValue = this.localStringValue.replace(
```
I think the spelling here is wrong.

Ahmed-Bhouri replied:
Thank you I fixed the typo

seanlip reviewed (changes requested):
Thanks, looks pretty good IMO. I think there's a spelling error though (replicated in multiple places)! PTAL.

Make sure to carefully test your changes after the fix.

seanlip commented:
Also @Ahmed-Bhouri maybe you should write some tests for this in schema-based-float-editor.component.spec.ts? (To lock in the correct behaviour and prevent it from breaking again)

Ahmed-Bhouri committed:
Merge remote-tracking branch 'upstream/develop' into inputmode-decimal

Ahmed-Bhouri committed:
Fixed typo

Ahmed-Bhouri committed:
Added tests

Ahmed-Bhouri commented:
@seanlip PTAL

I have added 2 tests to check for the value and fixed the typo

seanlip commented on core/templates/components/forms/schema-based-editors/schema-based-float-editor.component.spec.ts:
```diff
@@ -260,4 +260,16 @@ describe('Schema based float editor component', function () {
     expect(component.errorStringI18nKey).toEqual('Error');
   });
+
+  it('should not show error to diffrent decimal points', () => {
```
Spelling: different

(Also: to --> for)

Ahmed-Bhouri replied:
Thank you,
pushing changes rn

seanlip reviewed (changes requested):
Thanks @Ahmed-Bhouri. Just one more small thing, otherwise LGTM.

Also, please follow the instructions in our PR guidelines when requesting reviews -- you should reply to all comments before doing so. See step 5 [here](https://github.com/oppia/oppia/wiki/Make-a-pull-request#step-5-address-review-comments-until-all-reviewers-approve).

Ahmed-Bhouri committed:
Merge remote-tracking branch 'upstream/develop' into inputmode-decimal

Ahmed-Bhouri committed:
Fixed Typo

Ahmed-Bhouri commented:
@seanlip PTAL
I have addressed the issues

seanlip reviewed (approved):
Looks great. Thanks @Ahmed-Bhouri, and congrats on your first PR to Oppia! :tada:

seanlip commented:
Ah, this needs a frontend review. @Lawful2002 could you PTAL as codeowner? Thanks.

Lawful2002 reviewed (approved):
Thanks @Ahmed-Bhouri, LGTM!

seanlip commented:
@Ahmed-Bhouri The flake in the last merge attempt seems to be related to your changes. Could you PTAL and analyze the root cause so that we can be sure that this PR does not result in a flake in develop?

```
Chrome Headless 123.0.6312.58 (Linux x86_64) Schema based float editor component should not show error for different decimal points FAILED
	Error: Expected null to equal 12.5.
	    at <Jasmine>
	    at UserContext.<anonymous> (core/templates/combined-tests.spec.js:13368:38)
	    at ./node_modules/zone.js/dist/zone.js.ZoneDelegate.invoke (core/templates/combined-tests.spec.js:789599:30)
	    at ./node_modules/zone.js/dist/proxy.js.ProxyZoneSpec.onInvoke (core/templates/combined-tests.spec.js:789084:43)
	Error: Expected 'I18N_INTERACTIONS_NUMERIC_INPUT_NO_INVALID_CHARS' to equal null.
	    at <Jasmine>
	    at UserContext.<anonymous> (core/templates/combined-tests.spec.js:13369:46)
	    at ./node_modules/zone.js/dist/zone.js.ZoneDelegate.invoke (core/templates/combined-tests.spec.js:789599:30)
	    at ./node_modules/zone.js/dist/proxy.js.ProxyZoneSpec.onInvoke (core/templates/combined-tests.spec.js:789084:43)
	Error: Expected null to equal 12.5.
	    at <Jasmine>
	    at UserContext.<anonymous> (core/templates/combined-tests.spec.js:13372:38)
	    at ./node_modules/zone.js/dist/zone.js.ZoneDelegate.invoke (core/templates/combined-tests.spec.js:789599:30)
	    at ./node_modules/zone.js/dist/proxy.js.ProxyZoneSpec.onInvoke (core/templates/combined-tests.spec.js:789084:43)
	Error: Expected 'I18N_INTERACTIONS_NUMERIC_INPUT_NO_INVALID_CHARS' to equal null.
	    at <Jasmine>
	    at UserContext.<anonymous> (core/templates/combined-tests.spec.js:13373:46)
	    at ./node_modules/zone.js/dist/zone.js.ZoneDelegate.invoke (core/templates/combined-tests.spec.js:789599:30)
	    at ./node_modules/zone.js/dist/proxy.js.ProxyZoneSpec.onInvoke (core/templates/combined-tests.spec.js:789084:43)
Chrome Headless 123.0.6312.58 (Linux x86_64): Executed 319 of 8968 (1 FAILED) (0 secs / 2.741 secs)
Chrome Headless 123.0.6312.58 (Linux x86_64) Schema based float editor component should not show error for different decimal points FAILED
	Error: Expected null to equal 12.5.
	    at <Jasmine>
	    at UserContext.<anonymous> (core/templates/combined-tests.spec.js:13368:38)
	    at ./node_modules/zone.js/dist/zone.js.ZoneDelegate.invoke (core/templates/combined-tests.spec.js:789599:30)
	    at ./node_modules/zone.js/dist/proxy.js.ProxyZoneSpec.onInvoke (core/templates/combined-tests.spec.js:789084:43)
	Error: Expected 'I18N_INTERACTIONS_NUMERIC_INPUT_NO_INVALID_CHARS' to equal null.
	    at <Jasmine>
	    at UserContext.<anonymous> (core/templates/combined-tests.spec.js:13369:46)
	    at ./node_modules/zone.js/dist/zone.js.ZoneDelegate.invoke (core/templates/combined-tests.spec.js:789599:30)
	    at ./node_modules/zone.js/dist/proxy.js.ProxyZoneSpec.onInvoke (core/templates/combined-tests.spec.js:789084:43)
	Error: Expected null to equal 12.5.
	    at <Jasmine>
	    at UserContext.<anonymous> (core/templates/combined-tests.spec.js:13372:38)
	    at ./node_modules/zone.js/dist/zone.js.ZoneDelegate.invoke (core/templates/combined-tests.spec.js:789599:30)
	    at ./node_modules/zone.js/dist/proxy.js.ProxyZoneSpec.onInvoke (core/templates/combined-tests.spec.js:789084:43)
	Error: Expected 'I18N_INTERACTIONS_NUMERIC_INPUT_NO_INVALID_CHARS' to equal null.
	    at <Jasmine>
	    at UserContext.<anonymous> (core/templates/combined-tests.spec.js:13373:46)
	    at ./node_modules/zone.js/dist/zone.js.ZoneDelegate.invoke (core/templates/combined-tests.spec.js:789599:30)
	    at ./node_modules/zone.js/dist/proxy.js.ProxyZoneSpec.onInvoke (core/templates/combined-tests.spec.js:789084:43)
```

Ahmed-Bhouri committed:
Fixed Testing Flakes

Ahmed-Bhouri committed:
Merge remote-tracking branch 'upstream/develop' into inputmode-decimal

Ahmed-Bhouri commented:
@seanlip PTAL

To solve this I started by locking the test seed to 505 (the one that fails) to be able to debug the tests easily.
I discovered that my test needs more details to test all possible combinations so I fixed that. That made me understand the origin of the test failing. It turns out the flakes are caused by the NumericInputValidationService where any number with commas is considered to be a non valid number so I added it to the valid characters list solving the issue in the test flake and for all cases. 

I am pretty positive now that there will be no more flakes in testing.

Thank you,

seanlip reviewed (approved):
Thanks @Ahmed-Bhouri! LGTM :)

seanlip commented:
@lkbhitesh07 PTAL as codeowner. Thanks!

lkbhitesh07 reviewed (approved):
LGTM for the code owner file.
Thanks
